onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,205 @@
+  std::vector<int32_t> axes;
+  if (node_unit.SinceVersion() >= 18) {
+    if (inputs.size() > 1 && inputs[1].node_arg.Exists()) {
+      // ReduceMean-18 uses the second optional input as axes
+      const auto& initializers(model_builder.GetInitializerTensors());
+      const auto& axes_tensor = *initializers.at(inputs[1].node_arg.Name());
+      Initializer unpacked_tensor(axes_tensor);
+      auto raw_axes = unpacked_tensor.DataAsSpan<int64_t>();
+      axes = OnnxIndexesToNnapi(raw_axes, input_shape.size());
+    }
+  } else {
+    // For ReduceMean-13 or earlier, retrieve axes from the attribute
+    const auto& axes_int64 = helper.Get("axes", std::vector<int64_t>{});
+    axes = OnnxIndexesToNnapi(gsl::span<const int64_t>(axes_int64.data(), axes_int64.size()), input_shape.size());
+  }
+
+  if (axes.empty() && !noop_with_empty_axes) {
+    // we provide an input with all axes for NNAPI here to simulate this default behavior to reduce all dimensions
+    axes.resize(input_shape.size());
+    std::iota(axes.begin(), axes.end(), 0);
+  }
+
+  // Add ReduceMean operation
+  InlinedVector<uint32_t> input_indices;
+  input_indices.push_back(operand_indices.at(inputs[0].node_arg.Name()));  // data
+
+  const auto axes_name = model_builder.GetUniqueName(node_unit.Name() + inputs[0].node_arg.Name() + "_axes");
+  Shape axes_dimen = {static_cast<uint32_t>(axes.size())};
+  const OperandType axes_operand_type(Type::TENSOR_INT32, axes_dimen);
+  ORT_RETURN_IF_ERROR(model_builder.AddOperandFromPersistMemoryBuffer(axes_name, axes.data(), axes_operand_type));
+
+  input_indices.push_back(operand_indices.at(axes_name));  // axes
+
+  int32_t input_size = static_cast<int32_t>(input_shape.size());
+
+  // Make output dimensions
+  InlinedVector<uint32_t> output_dimen;
+  if (keepdims) {
+    output_dimen.reserve(input_size);
+  } else {
+    output_dimen.reserve(input_size - axes.size());
+  }
+
+  for (int32_t i = 0; i < input_size; i++) {
+    if (std::find(axes.begin(), axes.end(), i) == axes.end()) {
+      output_dimen.push_back(input_shape[i]);
+    } else {
+      if (keepdims) {
+        output_dimen.push_back(1);
+      }
+    }
+  }
+
+  // In case of a tensor has all 1's in dimension such as {1,1,1,1} and gets all reduced,
+  // NNAPI requires the output shape to be {1}. (otherwise NNAPI will treat it as dynamic shape.)
+  if (output_dimen.empty())
+    output_dimen.push_back(1);
+
+  shaper.AddShape(output, output_dimen);
+
+  ADD_SCALAR_OPERAND(model_builder, input_indices, helper.Get("keepdims", 1));
+
+  const OperandType output_operand_type(operand_types.at(inputs[0].node_arg.Name()).type, output_dimen);
+  ORT_RETURN_IF_ERROR(model_builder.AddOperation(op_code, input_indices,
+                                                 {output}, {output_operand_type}));
+
+  if (axes.empty() && noop_with_empty_axes) {
```

Comment:
I think this is correct, but a little harder to follow than before. e.g., it could be more obvious that we either add a mean or an identity operation.
```
// after modifying axes to account for !noop_with_empty_axes
if (!axes.empty()) {
  // add mean
} else {
  // add identity
}
```

Reply:
actually it probably should be updated slightly, currently it will add the mean operation even if axes is empty right?

Reply:
right!  actually I shouldn't remove the original `if (!axes.empty()){  }` block